src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+        storage_metadata: &CollectionMetadata,
+        now: NowFn,
+    ) -> anyhow::Result<Self> {
+        trace!("Initializing healthchecker for source {source_name}");
+        let mut persist_clients = persist_clients.lock().await;
+        let persist_client = persist_clients
+            .open(storage_metadata.persist_location.clone())
+            .await
+            .context("error creating persist client for Healthchecker")?;
+        drop(persist_clients);
+
+        let (write_handle, read_handle) = persist_client
+            .open(storage_metadata.status_shard)
+            .await
+            .context("error opening Healthchecker persist shard")?;
+
+        let (since, upper) = (read_handle.since().clone(), write_handle.upper().clone());
+
+        let listener = read_handle
+            .clone()
+            .await
+            .listen(since.clone())
+            .await
+            .expect("since <= as_of asserted");
+
+        let mut healthchecker = Self {
+            worker_id,
+            worker_count,
+            source_type,
+            source_name,
+            source_id: source_id.to_string(),
+            upstream_name,
+            current_status: SourceStatus::Starting,
+            active,
+            upper: Antichain::from_elem(Timestamp::minimum()),
+            write_handle,
+            listener,
+            now,
+        };
+
+        // Bootstrap should reload the previous state of the source, if any
+        healthchecker.bootstrap_state(read_handle, &upper).await;
+        tracing::trace!(
+            "Healthchecker for source {} at status {} finished bootstrapping!",
+            &healthchecker.source_name,
+            &healthchecker.current_status
+        );
+
+        Ok(healthchecker)
+    }
+
+    /// Advances the upper to ensure the collection is queryable, even when the status
+    /// is unchanged.
+    pub async fn advance_upper(&mut self) {
+        if self.active {
+            let updates: Vec<((SourceData, ()), u64, i64)> = vec![];
+            loop {
+                let next_ts = (self.now)();
+                let new_upper = Antichain::from_elem(next_ts + 1);
+                match self
+                    .write_handle
+                    .compare_and_append(updates.iter(), self.upper.clone(), new_upper.clone())
+                    .await
+                {
+                    Ok(Ok(Ok(()))) => {
+                        trace!("Upper advanced to {new_upper:?}!");
+                        self.upper = new_upper;
+                        break;
+                    }
+                    Ok(Ok(Err(_))) => {
+                        // Upper is already up to date, no need to do any work
+                        break;
+                    }
+                    Ok(Err(invalid_use)) => panic!("compare_and_append failed: {invalid_use}"),
+                    Err(external_err) => {
+                        panic!("compare_and_append failed: {external_err}");
```

Comment:
Ah, nice catch, the `loop` is there mostly because this was originally ~copy-and-pasted~ inspired by `update_status`. I also didn't pay enough attention to what this specific error meant, in my mind it was more of a "things are so out of control that it's better to panic and restart storaged" than "oh, S3 is just having a bad day and we got timed out".

Reply:
These errors can happen in situations like "your append was succesful but sth timed out and we didn't hear back so we don't know for sure if it was successful or not". And in cases where it's not harmful to try again and potentially advance the timestamp a bit further. Or in cases we sync up our state and then see if we still need to emit our status update it's good to retry.